### PR TITLE
Broke everything

### PR DESCRIPTION
**Pull request by msmith-techempower**
... into more manageable modules.

This was in an attempt to isolate docker to `docker_helper` so that I could implement the `--build` flag, but that quickly took me down the rabbit hole. Needs LOTS of QA.

Addresses [#27](https://github.com/TechEmpower/FrameworkBenchmarks/issues/3296)

### CODE REVIEW

**Comment by msmith-techempower** [comment]
Hahaha... WOO PASSING TRAVIS!!!

/checks_each_test

... each test is broken and the travis script is broken too, so everything fails with a pass.

✅

**Comment by nbrady-techempower** [comment]
I like what I'm seeing. 🥇

File: toolset/benchmark/test_types/db_type.py [hunk 1]
```diff
@@ -19,7 +19,8 @@ def get_url(self):
         return self.db_url
 
     def verify(self, base_url):
-        '''Ensures body is valid JSON with a key 'id' and a key 
+        '''
```
Comment (nbrady-techempower) [inline toolset/benchmark/test_types/db_type.py]: I'm not a fan of this style of comment, but it seems like we switch between this and `#`. Is there a reason/standard?

Comment (msmith-techempower) [reply toolset/benchmark/test_types/db_type.py]: This is the way python wants you to document your functions.

File: toolset/utils/docker_helper.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    #                     forcerm=True):
+    #             handle_build_output(line)
+    #     except Exception as e:
+    #         tee_output(prefix, "Docker build failed; terminating\n")
+    #         print(e)
+    #         return 1
+
+
+def find_docker_file(path, pattern):
+    '''
```
Comment (nbrady-techempower) [inline toolset/utils/docker_helper.py]: indentation madness

File: toolset/utils/docker_helper.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  the given path. 
+  '''
+    for root, dirs, files in os.walk(path):
+        for name in files:
+            if fnmatch.fnmatch(name, pattern):
+                return os.path.join(root, name)
+
+
+def gather_docker_dependencies(docker_file):
+    '''
```
Comment (nbrady-techempower) [inline toolset/utils/docker_helper.py]: more madness

File: toolset/benchmark/fortune_html_parser.py [hunk 4]
```diff
@@ -25,140 +26,146 @@ class FortuneHTMLParser(HTMLParser):
...
-    # The spec says that for HTML this is case insensitive,
-    # and since we did not specify xml compliance (where
-    # incorrect casing would throw a syntax error), we must
-    # allow all casings. We will lower for our normalization.
-    self.body.append("<!{d}>".format(d=decl.lower()))
+    # Is called when a doctype or other such tag is read in.
+    # For our purposes, we assume this is only going to be
+    # "DOCTYPE html", so we will surround it with "<!" and ">".
+    def handle_decl(self, decl):
+        # The spec says that for HTML this is case insensitive,
```
Comment (nbrady-techempower) [inline toolset/benchmark/fortune_html_parser.py]: So this should be a `'''` comment block?

File: toolset/utils/docker_helper.py (hunk 3 above)
Comment (msmith-techempower) [reply toolset/utils/docker_helper.py]: `docker_helper.py` is not ready yet

File: toolset/benchmark/fortune_html_parser.py (hunk 4 above)
Comment (msmith-techempower) [reply toolset/benchmark/fortune_html_parser.py]: Prolly

File: toolset/benchmark/test_types/verifications.py [hunk 5]
```diff
@@ -48,56 +48,49 @@ def verify_headers(headers, url, should_be='json'):
...
-    for v in (v for v in ('Server', 'Date', 'Content-Type') if v.lower() not in headers):
-        problems.append(
-            ('warn', 'Required response header missing: %s' % v, url))
-
-    if all(v.lower() not in headers for v in ('Content-Length', 'Transfer-Encoding')):
-        problems.append(
-            ('warn',
-             'Required response size header missing, please include either "Content-Length" or "Transfer-Encoding"',
-             url))
+    for v in (v for v in ('Server', 'Date', 'Content-Type')
```
Comment (nbrady-techempower) [inline toolset/benchmark/test_types/verifications.py]: I know you didnt add this but wow.. for v in v for v in.

File: toolset/setup/linux/prerequisites.sh [hunk 6]
```diff
@@ -39,10 +39,7 @@ sudo apt-get -qqy install -o Dpkg::Options::="--force-confdef" -o Dpkg::Options:
   #  mlton \
   # re2c libnuma-dev
```
Comment (nbrady-techempower) [inline toolset/setup/linux/prerequisites.sh]: You can remove all these comments now. They exist in the tfb base.

File: toolset/setup/linux/prerequisites.sh [hunk 7]
```diff
@@ -39,10 +39,7 @@ sudo apt-get -qqy install -o Dpkg::Options::="--force-confdef" -o Dpkg::Options:
...
 
-sudo pip install colorama==0.3.1
-# Version 2.3 has a nice Counter() and other features
-# but it requires --allow-external and --allow-unverified
-sudo pip install progressbar==2.2 requests MySQL-python psycopg2 pymongo
+sudo pip install colorama==0.3.1 requests MySQL-python psycopg2-binary pymongo
```
Comment (nbrady-techempower) [inline toolset/setup/linux/prerequisites.sh]: Yay! Fixes that annoying warning.

File: toolset/benchmark/framework_test.py [hunk 8]
```diff
@@ -1,1017 +1,204 @@
...
-from utils import header
-from utils import gather_docker_dependencies
-from utils import find_docker_file
+from toolset.utils.output_helper import header
+from toolset.utils import docker_helper
 
 # Cross-platform colored text
 from colorama import Fore, Back, Style
 from datetime import datetime
 from datetime import timedelta
```
Comment (nbrady-techempower) [inline toolset/benchmark/framework_test.py]: you can probably cut 10 more lines bc of unused imports here. importlib, socket, time ppring, json, multiprocessing, collections, threading, datetime

File: toolset/benchmark/test_types/framework_test_type.py [hunk 9]
```diff
@@ -146,7 +151,8 @@ def get_current_world_table(self):
 
         if database_name == "mysql":
             try:
-                db = MySQLdb.connect("TFB-database", "benchmarkdbuser", "benchmarkdbpass", "hello_world")
+                db = MySQLdb.connect("TFB-database", "benchmarkdbuser",
+                                     "benchmarkdbpass", "hello_world")
```
Comment (nbrady-techempower) [inline toolset/benchmark/test_types/framework_test_type.py]: Now might be a good time for a constants.py

File: toolset/benchmark/benchmarker.py [hunk 10]
```diff
@@ -1,687 +1,588 @@
...
 import os
-import uuid
-import shutil
-import stat
-import json
-import requests
 import subprocess
 import traceback
-import time
 import pprint
```
Comment (nbrady-techempower) [inline toolset/benchmark/benchmarker.py]: i believe this is unused as well

Comment (msmith-techempower) [reply toolset/benchmark/benchmarker.py]: The ol' import the package then import the function from that package you need.

File: toolset/benchmark/test_types/framework_test_type.py (hunk 9 above)
Comment (msmith-techempower) [reply toolset/benchmark/test_types/framework_test_type.py]: I tried this out and disliked it here. These values will never change and you'll never get a meaningful constant that is as short as these strings, magic though they may be.

**Pull request merged**